docs/howto/use_modules_and_layers.md

```diff
@@ -0,0 +1,392 @@
+
+In general we need more than just a DB connection though. We need components that enable us to perform different operations, and we need to be able to wire them together. This is what _modules_ are for.
+
+## Our first ZIO module
+The module pattern is _the way_ ZIO manages dependencies between application components.
+In its initial formulation it was only using trait mix-ins, as shown in [Appendix](#appendix-the-classic-module-formulation-until-version-100-rc17).
+
+Here we will see the new formulation of the module pattern, that resolves most of the shortcomings of the previous version.
+
+### What is a module?
+A module is a group of functions that deals with only one concern. Keeping the scope of a module limited improves our ability to understand code, in that we need to focus
+ only on one topic at a time without juggling with too many concepts together in our head.
+ 
+`ZIO` iself provides the basic capabilities through modules, e.g. see how `ZEnv` is defined.
+
+### The module recipe
+Let's build a module for user data access, following these simple steps:
+
+1. Define an object that gives the name to the module, this can be (not necessarily) a package object
+1. Within the module object define a `trait Service` that defines the interface our module is exposing, in our case 2 methods to retrieve a product details, and one to retrieve the related products
+1. Within the module object define a type alias like `type ModuleName = Has[Service]` (see below for details on `Has`)
+1. Within the module object define the different implementations of `ModuleName` through `ZLayer` (see below for details on `ZLayer`)
+
+```scala mdoc:silent
+import zio.{Has, ZLayer}
+
+object UserRepo {
+  trait Service {
+    def getProductDetails(productId: ProductId): IO[DBError, Product]
+    def getRelatedProducts(productId: ProductId): IO[DBError, List[Product]]
+  }
+  
+  type UserRepo = Has[Service]
+  val testRepo: ZLayer[Any, Nothing, UserRepo] = ???
+  
+}
+```
+
+We encountered two new data types `Has` and `ZLayer`, let's get familiar with them.
+
+### The `Has` data type
+
+`Has[A]` represents a dependency on a service of type `A`. Two `Has[_]` can be combined _horizontally_ through `+` and `++` operators, as in
+
+```scala mdoc:silent
+val repo: Has[Repo.Service] = ???
+val logger: Has[Logger.Service] = ???
+
+val mix: Has[Repo.Service] with Has[Logger.Service] = repo ++ logger
+``` 
+
+At this point you might ask: what's the use of `Has` if the resulting type is still a mix of two traits? The extra power given by `Has` is that the resulting data structure is backed by an _etherogeneus map_ from service type to service implementation, that collects each instance that is mixed in so that the instances can be accessed/extracted/modified individually, still guaranteeing supreme type safety.
+
+```scala mdoc:silent
+// get back the repo from the mixed value:
+val log = mix.get[Has[Logger.Service]].log("Hello modules!")
```

Comment:
`val log = mix.get[Logger.Service].log("Hello modules!")`

the `Has[A]` is the one that provides the `get[A]` operation